native/src/main/java/io/ballerina/stdlib/task/utils/TaskJob.java

```diff
@@ -35,24 +48,66 @@ public class TaskJob implements Job {
         Thread.startVirtualThread(() -> {
             Runtime runtime = TaskManager.getInstance().getRuntime();
+            Boolean isTokenHolder = (Boolean) jobExecutionContext.getMergedJobDataMap().get(TOKEN_HOLDER);
             BObject job = (BObject) jobExecutionContext.getMergedJobDataMap().get(TaskConstants.JOB);
-            try {
-                ObjectType objectType = (ObjectType) job.getOriginalType();
-                boolean isConcurrentSafe = objectType.isIsolated() && objectType.isIsolated(TaskConstants.EXECUTE);
-                StrandMetadata metadata = new StrandMetadata(isConcurrentSafe, null);
-                runtime.callMethod(job, TaskConstants.EXECUTE, metadata);
-            } catch (BError error) {
-                Utils.notifyFailure(jobExecutionContext, error);
-            } catch (Throwable t) {
-                Utils.notifyFailure(jobExecutionContext, ErrorCreator.createError(t));
+            boolean needsRollback = false;
+            if (isTokenHolder == null || isTokenHolder) {
+                try {
+                    ObjectType objectType = (ObjectType) job.getOriginalType();
+                    boolean isConcurrentSafe = objectType.isIsolated() && objectType.isIsolated(TaskConstants.EXECUTE);
+                    StrandMetadata metadata = new StrandMetadata(isConcurrentSafe, null);
+                    runtime.callMethod(job, TaskConstants.EXECUTE, metadata);
+                } catch (BError error) {
+                    Utils.notifyFailure(jobExecutionContext, error);
+                } catch (Throwable t) {
+                    Utils.notifyFailure(jobExecutionContext, ErrorCreator.createError(t));
+                }
+            } else {
+                Connection connection = (Connection) jobExecutionContext.getMergedJobDataMap().get(CONNECTION);
+                String instanceId = ((BString) jobExecutionContext.getMergedJobDataMap().get(INSTANCE_ID)).getValue();
+                int livenessInterval = (int) jobExecutionContext.getMergedJobDataMap().get(LIVENESS_INTERVAL);
+                try {
+                    connection.setAutoCommit(false);
+                    needsRollback = true;
+                    boolean acquired
+                            = attemptTokenAcquisition(connection, instanceId, false, livenessInterval);
+                    connection.commit();
+                    if (acquired) {
+                        ObjectType objectType = (ObjectType) job.getOriginalType();
+                        boolean isConcurrentSafe = objectType.isIsolated()
+                                && objectType.isIsolated(TaskConstants.EXECUTE);
+                        StrandMetadata metadata = new StrandMetadata(isConcurrentSafe, null);
+                        runtime.callMethod(job, TaskConstants.EXECUTE, metadata);
+                    } else {
+                        upsertToken(connection, instanceId);
+                    }
+                    needsRollback = false;
```

Comment:
Use a try block that performs rollback automatically if commit is not reached. That way, you can eliminate the needsRollback flag. 
```java
try {
    connection.setAutoCommit(false);
    boolean acquired = attemptTokenAcquisition(...);
    connection.commit();
    ...
} catch (...) {
    rollback(connection);
}
```